segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentContainerExtension.java

```diff
@@ -0,0 +1,35 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.server;
+
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Defines an Extension that can be associated with a {@link SegmentContainer}, which will be instantiated (and initialized)
+ * when the {@link SegmentContainer} is initialized, and closed when the {@link SegmentContainer} is closed.
+ */
+public interface SegmentContainerExtension extends AutoCloseable {
+    @Override
+    void close();
+
+    /**
+     * Performs any initialization required for this SegmentContainerExtension instance.
+     * @return A CompletableFuture that, when completed, will indicate that the initialization is done.
+     */
+    CompletableFuture<Void> initialize();
+
+    /**
+     * Creates a Collection of any additional WriterSegmentProcessors to pass to the StorageWriter.
+     * @param metadata The Metadata of the Segment to create Writer Processors for.
+     * @return A Collection containing the desired processors. If none are required, this should return an empty list.
+     */
+    Collection<WriterSegmentProcessor> createWriterSegmentProcessors(UpdateableSegmentMetadata metadata);
```

Comment:
I need a way to inject "extensions" into the `StorageWriter` as well (at the segment level). These extensions are associated with the `SegmentContainerExtension` that owns them and has the context required to create them. 

The alternative to the presented approach would be to have another type of extension that is passed to the `StorageWriterFactory` (different from `SegmentContainerFactory`) so that it can create that. That will not only lead to a lot more code and possibilities of messing up (like setting up one but forgetting the other one), but I will not be able to communicate between the `SegmentContainerExtension` and the `WriterTableProcessor`s associated with it. For Tables, I need the former to know when the latter has finished post-recovery indexing so that it may serve conditional updates.